File: swift/004.swift
// https://projecteuler.net/problem=4

import Darwin

func largestPalindromeProduct(digits: Int) -> Int {
  let largestFactor = Int(pow(10.0, Float(digits))) - 1
  let smallestFactor = Int(pow(10.0, Float(digits - 1)))
  var largestProduct = 0

  for i in reverse(smallestFactor...largestFactor) {
    // If i is less than the square root of the largest product, we have already
    // found the largest product
    if i * i < largestProduct {
      break
    }

    for j in reverse(smallestFactor...i) {
      var product = i * j
      var productString = String(product)

      if String(reverse(productString)) == productString && product > largestProduct {
        largestProduct = product
        // Continue to next iteration of outer loop - no larger products can be
        // found within this iteration
        continue
      }
    }
  }

  return largestProduct
}

println(largestPalindromeProduct(3))

File: elixir/004.exs
# https://projecteuler.net/problem=4

defmodule PalindromicProduct do
  def largest(digits) do
    largest_multiplier = round(:math.pow(10, digits)) - 1
    smallest_multiplier = round(:math.pow(10, digits - 1))

    _largest(smallest_multiplier, largest_multiplier, largest_multiplier, 0)
  end

  def _largest(_, a, b, largest_product)
    when a < 10 and b < 10 do
    largest_product
  end

  def _largest(smallest_multiplier, a, b, largest_product)
    when b < smallest_multiplier do
    _largest(smallest_multiplier, a - 1, a - 1, largest_product)
  end

  def _largest(smallest_multiplier, a, b, largest_product) do
    product_string = Integer.to_string(a * b)

    if product_string === String.reverse(product_string) do
      _largest(smallest_multiplier, a, b - 1, Kernel.max(largest_product, a * b))
    else
      _largest(smallest_multiplier, a, b - 1, largest_product)
    end
  end
end

IO.inspect(PalindromicProduct.largest(3))
